# Roll number: 20BM6JP15
# Name: Utkarsh Yadav
# Assignment number: 3

# Gradient Boosting

Problem Statement: Diabetes classification using XGBoost Classifier.
Dataset: Please download the dataset from the link:
https://www.dropbox.com/s/c1q3qix77hclbf2/diabetes.csv?dl=0
The dataset consists of 9 features like “Glucose”, “BloodPressure” etc. The target variable is
the field named as “Outcome” which holds 2 values, 0 or 1.

In [1]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt

from sklearn.model_selection import train_test_split

from xgboost import XGBClassifier
from sklearn import metrics

### Loading data

In [2]:
data = pd.read_csv("diabetes.csv")

In [3]:
data

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

In [4]:
X = data.copy()
X = X.drop(['Outcome'],axis=1)
X

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  
0                       0.627   50  
1                       0.351   31  
2                       0.672   32  
3                       0.167   21  
4                       2.288   33  
..                        ...  ...  
763                     0.171   63  
764                     0.340   27  
765                     0.245   30  
766                     0.349   47  
767                     0.315   23  

[768 rows x 8 columns]

In [5]:
y = data['Outcome']
y

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64

# Use a 80-20 train-test split for the experiment.

In [6]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

# Assignment Tasks: In this assignment, please use the XGBClassifier available under XGBoost python package to perform classification on the Diabetes dataset. Please vary the following parameters and report the test set accuracies for each combination below:

# 1. Learning rate = 0.1, objective = logistic regression

In [7]:
xgb = XGBClassifier(learning_rate=0.1, objective="binary:logistic" , eval_metric = 'logloss',use_label_encoder=False)
xgb.fit(X_train,y_train)

print("\nAccuracy with Learning rate = 0.1, objective = logistic regression is :\n",xgb.score(X_test,y_test))


Accuracy with Learning rate = 0.1, objective = logistic regression is :
 0.7727272727272727


# 2. Learning rate = 0.1, objective = hinge loss

In [8]:
xgb = XGBClassifier(learning_rate=0.1, objective="binary:hinge", eval_metric = 'logloss',use_label_encoder=False)
xgb.fit(X_train,y_train)

print("\nAccuracy with Learning rate = 0.1, objective = hinge loss is :\n",xgb.score(X_test,y_test))


Accuracy with Learning rate = 0.1, objective = hinge loss is :
 0.7337662337662337


# 3. Learning rate = 0.3, objective = logistic regression, max_depth = 2

In [9]:
xgb = XGBClassifier(learning_rate=0.3, objective="binary:logistic", max_depth = 2, eval_metric = 'logloss',use_label_encoder=False)
xgb.fit(X_train,y_train)

print("\nAccuracy with Learning rate = 0.3, objective = logistic regression, max_depth = 2 is :\n",xgb.score(X_test,y_test))


Accuracy with Learning rate = 0.3, objective = logistic regression, max_depth = 2 is :
 0.7922077922077922


# 4. Learning rate = 0.3, objective = logistic regression, max_depth = 8

In [10]:
xgb = XGBClassifier(learning_rate=0.3, objective="binary:logistic", max_depth = 8, eval_metric = 'logloss',use_label_encoder=False)
xgb.fit(X_train,y_train)

print("\nAccuracy with Learning rate = 0.3, objective = logistic regression, max_depth = 8 is :\n",xgb.score(X_test,y_test))


Accuracy with Learning rate = 0.3, objective = logistic regression, max_depth = 8 is :
 0.7337662337662337


# 5. Learning rate = 0.7, objective = logistic regression

In [11]:
xgb = XGBClassifier(learning_rate=0.7, objective="binary:logistic", eval_metric = 'logloss',use_label_encoder=False)
xgb.fit(X_train,y_train)

print("\nAccuracy with Learning rate = 0.7, objective = logistic regression is :\n",xgb.score(X_test,y_test))


Accuracy with Learning rate = 0.7, objective = logistic regression is :
 0.7402597402597403


# 6. Learning rate = 0.7, objective = hinge loss

In [12]:
xgb = XGBClassifier(learning_rate=0.7, objective="binary:hinge", eval_metric = 'logloss',use_label_encoder=False)
xgb.fit(X_train,y_train)

print("\nAccuracy with Learning rate = 0.7, objective = hinge loss is :\n",xgb.score(X_test,y_test))


Accuracy with Learning rate = 0.7, objective = hinge loss is :
 0.7337662337662337


# 7. Learning rate = 0.7, objective = hinge loss, max_depth = 8

In [13]:
xgb = XGBClassifier(learning_rate=0.7, objective="binary:hinge", max_depth = 8, eval_metric = 'logloss',use_label_encoder=False)
xgb.fit(X_train,y_train)

print("\nAccuracy with Learning rate = 0.7, objective = hinge loss, max_depth = 8 is :\n",xgb.score(X_test,y_test))


Accuracy with Learning rate = 0.7, objective = hinge loss, max_depth = 8 is :
 0.7077922077922078


# 8. Learning rate = 0.3, objective = logistic regression, L1 regularisation = 0.2, max_depth = 8

In [14]:
xgb = XGBClassifier(learning_rate=0.3, objective="binary:logistic",reg_lambda=0.2, max_depth = 8, eval_metric = 'logloss',use_label_encoder=False)
xgb.fit(X_train,y_train)

print("\nAccuracy with Learning rate = 0.3, objective = logistic regression, L1 regularisation = 0.2, max_depth = 8 is :\n",xgb.score(X_test,y_test))


Accuracy with Learning rate = 0.3, objective = logistic regression, L1 regularisation = 0.2, max_depth = 8 is :
 0.7857142857142857


# 9. Learning rate = 0.3, objective = logistic regression, L2 regularisation = 0.2, max_depth= 8

In [15]:
xgb = XGBClassifier(learning_rate=0.3, objective="binary:logistic",reg_alpha=0.2, max_depth = 8, eval_metric = 'logloss',use_label_encoder=False)
xgb.fit(X_train,y_train)

print("\nAccuracy with Learning rate = 0.3, objective = logistic regression, L2 regularisation = 0.2, max_depth= 8 is :\n",xgb.score(X_test,y_test))


Accuracy with Learning rate = 0.3, objective = logistic regression, L2 regularisation = 0.2, max_depth= 8 is :
 0.7727272727272727


# 10. Learning rate = 0.3, objective = logistic regression, split finding algorithm = Approximation Algorithm (present in the original paper https://arxiv.org/pdf/1603.02754.pdf )

In [16]:
# https://xgboost.readthedocs.io/en/latest/treemethod.html

xgb = XGBClassifier(learning_rate=0.3, objective="binary:logistic",tree_method="approx", eval_metric = 'logloss',use_label_encoder=False)
xgb.fit(X_train,y_train)

print("\nAccuracy with Learning rate = 0.3, objective = logistic regression, split finding algorithm = Approximation Algorithm (present in the original paper https://arxiv.org/pdf/1603.02754.pdf ) is :\n",xgb.score(X_test,y_test))


Accuracy with Learning rate = 0.3, objective = logistic regression, split finding algorithm = Approximation Algorithm (present in the original paper https://arxiv.org/pdf/1603.02754.pdf ) is :
 0.7727272727272727


# Summary 

Accuracy with Learning rate = 0.1, objective = logistic regression is : 0.7727272727272727
    
Accuracy with Learning rate = 0.1, objective = hinge loss is : 0.7337662337662337
    
Accuracy with Learning rate = 0.3, objective = logistic regression, max_depth = 2 is : 0.7922077922077922
    
Accuracy with Learning rate = 0.3, objective = logistic regression, max_depth = 8 is : 0.7337662337662337
    
Accuracy with Learning rate = 0.7, objective = logistic regression is : 0.7402597402597403
    
Accuracy with Learning rate = 0.7, objective = hinge loss is : 0.7337662337662337
    
Accuracy with Learning rate = 0.7, objective = hinge loss, max_depth = 8 is : 0.7077922077922078
    
Accuracy with Learning rate = 0.3, objective = logistic regression, L1 regularisation = 0.2, max_depth = 8 is: 0.7857142857142857
    
Accuracy with Learning rate = 0.3, objective = logistic regression, L2 regularisation = 0.2, max_depth= 8 is : 0.7727272727272727
    
Accuracy with Learning rate = 0.3, objective = logistic regression, split finding algorithm = Approximation Algorithm (present in the original paper https://arxiv.org/pdf/1603.02754.pdf ) is : 0.7727272727272727